zerver/lib/actions.py

```diff
@@ -7129,8 +7129,17 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
             # avoid sending two events with new muted topics in
             # immediate succession; this is correct only because
             # muted_topics events always send the full set of topics.
-            remove_topic_mute(muting_user, stream_being_edited.id, orig_topic_name)
-            do_mute_topic(muting_user, stream_being_edited, topic_name)
+            if topic_name is not None and new_stream is None:
+                remove_topic_mute(muting_user, stream_being_edited.id, orig_topic_name)
+                do_mute_topic(muting_user, stream_being_edited, topic_name)
+            elif new_stream is not None:
+                if muting_user.id not in delete_event_notify_user_ids:
+                    remove_topic_mute(muting_user, stream_being_edited.id, orig_topic_name)
+                    do_mute_topic(
+                        muting_user,
+                        new_stream,
+                        topic_name if topic_name is not None else orig_topic_name,
+                    )
```

Comment:
Thinking about this more, I think in the fallthrough case (i..e you are in `delete_event_notify_user_ids`), we should unmute the topic for the user, since for you, conceptually, the topic is being deleted.